optimum/habana/transformers/trainer.py

```diff
@@ -414,6 +415,50 @@ def _wrap_model(self, model, training=True, dataloader=None):
 
         return model
 
+    def _maybe_log_save_evaluate(self, tr_loss, model, trial, epoch, ignore_keys_for_eval):
+        if self.control.should_log:
+            if is_torch_tpu_available():
+                xm.mark_step()
+
+            logs: Dict[str, float] = {}
+
+            # all_gather + mean() to get average loss over all processes
+            tr_loss_scalar = self._nested_gather(tr_loss).mean().item()
+
+            # reset tr_loss to zero
+            tr_loss -= tr_loss
+            logs["loss"] = round(tr_loss_scalar / (self.state.global_step - self._globalstep_last_logged), 4)
+            logs["learning_rate"] = self._get_learning_rate()
+
+            self._total_loss_scalar += tr_loss_scalar
+            self._globalstep_last_logged = self.state.global_step
+            self.store_flos()
+
+            self.log(logs)
+
+        metrics = None
+        if self.control.should_evaluate:
+            if isinstance(self.eval_dataset, dict):
+                metrics = {}
+                for eval_dataset_name, eval_dataset in self.eval_dataset.items():
+                    dataset_metrics = self.evaluate(
+                        eval_dataset=eval_dataset,
+                        ignore_keys=ignore_keys_for_eval,
+                        metric_key_prefix=f"eval_{eval_dataset_name}",
+                    )
+                    metrics.update(dataset_metrics)
+            else:
+                metrics = self.evaluate(ignore_keys=ignore_keys_for_eval)
+            self._report_to_hp_search(trial, self.state.global_step, metrics)
+
+        if self.control.should_save:
+            if self.args.throughput_rm_save_ckpt_time:
+                save_start = time.time()
```

Comment:
Actually I think we should also measure the time taken for logging and evaluating right above in this method.
Maybe we can just measure the time at the begining and at the end of this method, WDYT?

Reply:
will that be too much difference with transformers? also there has loss over all ranks for 8x.

Reply:
I think it's fine since we are not modifying the base behavior but only adding a new argument to enable another behavior.

Reply:
You mean that the loss is logged by all processes in distributed training?

Reply:
right, loss is averaged across all ranks/processes in distributed training.

Reply:
Is this an issue?

Reply:
I think maybe we should be count that portion as it's part of 8x training.

Reply:
Anyway even in 8x training, the throughput that is logged is the throughput of the main process. So not sure that changes anything.